src/placeos-frontend-loader/api/remote/remote.cr

```diff
@@ -0,0 +1,150 @@
+    end
+
+    struct Commit
+      include JSON::Serializable
+      getter commit : String
+      getter date : String
+      getter author : String
+      getter subject : String
+
+      def initialize(@commit, @date, @author, @subject)
+      end
+    end
+
+    struct Reference
+      include JSON::Serializable
+
+      enum Type
+        GitLab
+        Github
+      end
+
+      getter repo_name : String
+      getter remote_type : Reference::Type
+      getter branch : String
+      getter tag : String | Nil
+
+      def initialize(@url : String, @branch : String? = "master", @tag : String? = nil, @hash : String? = "HEAD")
+        @repo_name = @url.split(".com/").last.rstrip("/")
+
+        if @url.includes?("github")
+          @remote_type = Type::Github
+        elsif @url.includes?("gitlab")
+          @remote_type = Type::GitLab
+        else
+          raise Exception.new("URL not supported")
+        end
+      end
+    end
+
+    abstract def commits(repo : String, branch : String) : Array(Commit)
+
+    abstract def branches(repo : String) : Hash(String, String)
+
+    abstract def releases(repo : String) : Array(String)
+
+    abstract def download(ref : Reference, path : String, branch : String? = "master", hash : String? = "HEAD", tag : String? = nil)
+
+    def extract_archive(dest_path : String, temp_tar_name : String)
+      raise File::NotFoundError.new(message: "File #{temp_tar_name} does not exist", file: temp_tar_name) unless File.exists?(Path.new(temp_tar_name))
+      if !Dir.exists?(Path.new(["./", dest_path]))
+        File.open(temp_tar_name) do |file|
+          begin
+            Compress::Gzip::Reader.open(file) do |gzip|
+              Crystar::Reader.open(gzip) do |tar|
+                tar.each_entry do |entry|
+                  next if entry.file_info.directory?
+                  parts = Path.new(entry.name).parts
+                  parts = parts.last(parts.size > 1 ? parts.size - 1 : 0)
+                  next if parts.size == 0
+                  file_path = Path.new([dest_path] + parts)
+                  Dir.mkdir_p(file_path.dirname) unless Dir.exists?(file_path.dirname)
+                  File.write(file_path, entry.io, perm: entry.file_info.permissions)
+                end
+              end
+            end
+          rescue ex : File::Error | Compress::Gzip::Error
+            Log.error(exception: ex) { "Could not unzip tar" }
+          end
+        end
+      end
+      File.delete(temp_tar_name)
+    end
+
+    def get_hash(hash : String, repository_uri : String, tag : String?, branch : String)
+      begin
+        if hash == "HEAD"
+          if (!tag.nil?)
+            hash = get_hash_by_tag(repository_uri, tag)
+          else
+            hash = get_hash_by_branch(repository_uri, branch)
+          end
+        end
+      rescue ex : KeyError
+        hash = get_hash_head(repository_uri) if hash.nil?
+      end
+      hash.not_nil!
+    end
```

Comment:
```suggestion
    def get_hash(hash : String, repository_uri : String, tag : String?, branch : String)
      if hash == "HEAD"
        if tag.nil?
          get_hash_by_branch(repository_uri, branch)
        else
          get_hash_by_tag(repository_uri, tag)
        end
      end
    rescue ex : KeyError
      get_hash_head(repository_uri)
    end
```

Reply:
Could you pls throw a comment on this saying that it's resolving a reference to a hash, as `get_hash` doesn't communicate this

Reply:
Updated my suggestion, realised I was missing a branch